lib/ask/runtime/channel_broker_state.ex

```diff
@@ -284,45 +295,64 @@ defmodule Ask.Runtime.ChannelBrokerState do
   #
   # FIXME: understand why Surveda would know about a contact having failed but
   #        the channel broker wouldn't have been notified?!
-  def clean_inactive_respondents(state, active_respondents) do
-    # TODO: Elixir 1.13 has Map.filter/2
-    new_active_contacts =
-      :maps.filter(
-        fn respondent_id, _ -> respondent_id in active_respondents end,
-        state.active_contacts
-      )
-
-    Map.put(state, :active_contacts, new_active_contacts)
+  def clean_inactive_respondents(state, active_respondent_ids) do
+    Repo.delete_all(
+      from q in Queue,
+        where:
+          q.channel_id == ^state.channel_id and not (q.respondent_id in ^active_respondent_ids) and
+            not is_nil(q.last_contact)
+    )
+
+    state
   end
 
-  # For leftover active contacts, we ask the remote channel for the actual IVR
-  # call or SMS message state. Keep only the contacts that are active or queued.
+  # For idle contacts, we ask the remote channel for the actual IVR call or SMS
+  # message state. Keep only the contacts that are active or queued.
+  #
+  # FIXME: this may take a while, and during that time the channel broker
+  #        won't process its mailbox, if it ever becomes a problem, we might
+  #        consider:
+  #
+  #        - only process a random N number of idle contacts on each call
+  #        - run the task in its own concurrent process
   def clean_outdated_respondents(state) do
     idle_time = gc_allowed_idle_time(state)
-    now = SystemTime.time().now
-
-    # TODO: Elixir 1.13 has Map.filter/2
-    new_active_contacts =
-      :maps.filter(
-        fn _, %{last_contact: last_contact} = active_contact ->
-          if DateTime.diff(now, last_contact, :second) < idle_time do
-            true
-          else
-            !Ask.Runtime.Channel.message_inactive?(state.runtime_channel, active_contact.channel_state)
-          end
-        end,
-        state.active_contacts
-      )
+    last_contact = SystemTime.time().now |> DateTime.add(-idle_time, :second)
+
+    query =
+      from q in Queue,
+        select: [:channel_id, :respondent_id, :channel_state],
+        where: q.channel_id == ^state.channel_id and q.last_contact < ^last_contact
+
+    Repo.all(query)
+    |> Enum.each(fn active_contact ->
+      if Ask.Runtime.Channel.message_inactive?(
+           state.runtime_channel,
+           active_contact.channel_state
+         ) do
+        Repo.delete(active_contact)
+      end
+    end)
```

Comment:
Note: we can't stream from the database as Ecto streams require to live inside a transaction, and we don't want to keep another long transaction open while we make external calls to verboice/nuntium, and we don't want to revert all the deletions because of a network error.

We could however load in batches from the database.